gnovm/pkg/gnolang/static_analysis.go

```diff
@@ -0,0 +1,268 @@
+package gnolang
+
+import (
+	"errors"
+	"fmt"
+)
+
+type StaticAnalysis struct {
+	// contexts for switch and for
+	contexts []Context
+	// funcContexts for functions and lambdas
+	funcContexts []FuncContext
+
+	// here we accumulate errors
+	// from lambdas defined in the function declaration
+	lambdaErrs []error
+}
+
+func NewStaticAnalysis() *StaticAnalysis {
+	return &StaticAnalysis{contexts: make([]Context, 0), funcContexts: make([]FuncContext, 0), lambdaErrs: make([]error, 0)}
+}
+
+func (s *StaticAnalysis) pushContext(ctx Context) {
+	s.contexts = append(s.contexts, ctx)
+}
+
+func (s *StaticAnalysis) popContext() Context {
+	last := s.contexts[len(s.contexts)-1]
+	s.contexts = s.contexts[0 : len(s.contexts)-1]
+	return last
+}
+
+func (s *StaticAnalysis) pushFuncContext(fc FuncContext) {
+	s.funcContexts = append(s.funcContexts, fc)
+}
+
+func (s *StaticAnalysis) popFuncContext() FuncContext {
+	last := s.funcContexts[len(s.funcContexts)-1]
+	s.funcContexts = s.funcContexts[0 : len(s.funcContexts)-1]
+	return last
+}
+
+// findCtxByLabel returns the last context if the label is empty
+// otherwise it returns the context that matches the label
+// if it doesn't exist, it returns nil
+func (s *StaticAnalysis) findCtxByLabel(label string) Context {
+	if len(label) == 0 {
+		return s.contexts[len(s.contexts)-1]
+	}
+
+	for i := len(s.contexts) - 1; i > -1; i-- {
+		if s.contexts[i].label() == label {
+			return s.contexts[i]
+		}
+	}
+
+	return nil
+}
+
+func (s *StaticAnalysis) Analyse(f *FuncDecl) []error {
+	s.pushFuncContext(&FuncDeclContext{
+		hasRet: false,
+		f:      f,
+	})
+	term := s.staticAnalysisBlockStmt(f.Body)
+
+	// todo use later maybe?
+	_ = s.popFuncContext().(*FuncDeclContext)
+
+	errs := make([]error, 0)
```

Comment:
I can't recall the global strategy in the repository. We use `go.uber.org/multierr` and `[]error` interchangeably. Could someone verify the primary usage and comment here if this PR should switch to the other method?

Reply:
By the way, the current implementation can only return 0 or 1 error. Is the []error for future needs? What would they be? In any case, multierr resolves this.

Reply:
we use `go.uber.org/multierr` only at one place